Fix extension management enable / save loading states

davecreaser opened the pull request:
## Description

- Attempt number two at fixing the super awkward and spaghetti like extension management loading states! 🎉 

Our talented QA engineer spotted some bugs with my previous fix - mostly to do with the `Save Changes` button when saving the settings for an extension. Also when testing with Metamask, which highlighted some other issues.

## Testing

⚠️  You'll need to login as Leela using Metamask. If you need help with that, let me know. Alternatively, log in as a random dev user with Metamask, but ensure they have permissions to manage extensions in a colony.

* Install the multisig extension.
* Change the threshold, and while that form is submitting, and the changes are pending, check that the `deprecate extension` button is disabled.
* Click on a different threshold, but **don't** click save changes (just ensure that the save changes button shows). Instead, click on `deprecate extension`. The `save changes` button should be disabled while this is loading.
* Change the threshold, and while that form is submitting, and the changes are pending, check that both the `uninstall extension` button and the `enable` button are disabled.
* Click on a different threshold, but **don't** click save changes. Instead, click on `enable`. The `save changes` button and the `uninstall extension` buttons should be disabled while this is loading.
* Deprecate the extension, then again choose a different threshold without saving (so that the save changes button appears). This time, uninstall the extension. The other buttons should be disabled while this happens.

Here's a video of me going through this flow, hope it helps!

https://github.com/user-attachments/assets/b7b638ac-718e-4136-a57f-332643bf530d

## Diffs

**Changes** 🏗

* Changes to the extension management button loading states

Resolves #4089

mmioana reviewed (dismissed):
Nice work on this issue @davecreaser 🙌 

It all works as expected 💯 

https://github.com/user-attachments/assets/5131a300-e3b8-4614-8123-35bdbcfe77c5

I'm wondering though if we could somehow unify the usage of `isPendingManagement || waitingForActionConfirmation` when disabling the buttons, given we use this condition almost everywhere. What do you think?

I'll approve this PR nonetheless 👍

iamsamgibbs reviewed (dismissed):
Nice job navigating through the extensions spaghetti, this is working nicely!

The only extra improvement I'd like to suggest is disabling the form itself as well as the buttons whilst each of these transactions are processing, as I show in some of my videos, I'm able to change the threshold whilst the transactions are processing. Otherwise all working great though!

Deprecate is disabled whilst saving changes:
https://github.com/user-attachments/assets/47d16334-3e28-4abd-89bf-e0631bfd8618
 
Save changes is disabled whilst depreacting:
https://github.com/user-attachments/assets/41af0c20-f51d-47a6-a5ea-b63de3306658

Save changes and uninstall are disabled whilst enabling:
https://github.com/user-attachments/assets/8de3235e-3721-486f-b8d3-4e274aa19829

Enable and save changes are disabled whilst uninstalling:
https://github.com/user-attachments/assets/b2e433fc-918c-4e68-b5e3-776d343760ab

davecreaser commented:
Thanks for the reviews @mmioana and @iamsamgibbs - I totally agree with both of your improvement points too so I've updated the PR with them. I've removed all references of waitingForActionConfirmation since it's now no longer needed, and the form should also be disabled along with the buttons. It did of course wipe your approvals though 🙃

iamsamgibbs reviewed (changes requested):
Perfect! Works very nicely now for multi-sig:

https://github.com/user-attachments/assets/200ca291-0aab-4724-9e1d-6a00ba70e92c

Can you apply the same changes to the staking advanced payments extension and reputation weighted extension? (If this is better to do as a new issue I can switch this to an approval!)

https://github.com/user-attachments/assets/de51feda-56b4-497d-9f4f-1397d9d3f54c

https://github.com/user-attachments/assets/77bcf665-25f5-4e55-9293-f8680cee8883

I noticed another issue too, seems to be unrelated to this PR as it is also happening on master, but possibly related to your previous extensions PR? The "Save changes" button gets stuck in a pending state if you reject the transaction for staking advanced payments. Multi-sig doesn't seem to have the same issue.

davecreaser commented:
> Can you apply the same changes to the staking advanced payments extension and reputation weighted extension? (If this is better to do as a new issue I can switch this to an approval!)

Oh this is totally an embarrassing goof on my part 🤦‍♂️ I had written a note to myself to cover those forms too and then got distracted by the dog and forgot all about them! Sorry to make you do an extra review @iamsamgibbs - fixing now!

davecreaser commented:
Okay, should be ready now! 🤞

mmioana reviewed (dismissed):
Really awesome work @davecreaser 👏 

Seems all states for the extensions are working properly now 🎉 

`Multi-sig Permissions`

https://github.com/user-attachments/assets/200d1bb4-0fb3-46c9-9693-73b0546a0af3

`Reputation Weighted`

https://github.com/user-attachments/assets/44bde723-70c9-4e02-a7ea-c54195457149

`Staking Advanced Payments`

https://github.com/user-attachments/assets/e169ccbc-bddc-4ef0-a0aa-a0b00989e4b9

iamsamgibbs reviewed (changes requested):
Thanks for the updates, this is getting closer!

Just the "hours" fields missed on reputation I'm afraid!

https://github.com/user-attachments/assets/b089f7af-24a5-46e8-a2b7-ca6305e83a90

Staking expenditures extension was absolutely fine, thanks for fixing the pending button issue too!

https://github.com/user-attachments/assets/0f2da5d2-4b3d-4354-aa48-7ed05d1829c0

Also had this weird issue with installing multi-sig, but haven't been able to recreate it. (I also should have checked the console before I stopped recording 🤦‍♂️)

https://github.com/user-attachments/assets/c7f91b90-1c95-4adb-81fd-80273cef7e34

davecreaser commented:
Agh, classic case of expecting the prop to be passed as expected, and not checking properly... 🙄 should be good now!

mmioana reviewed (approved):
I have completely missed testing the `Custom` settings for the `Reputation Weighted`, amazing eyes @iamsamgibbs 👀 

With your latest changes @davecreaser, all inputs for the `Custom` settings are disabled 🥇 

https://github.com/user-attachments/assets/6eeff5db-57cc-48e5-a404-f78ea6e1d71a

Amazing work!!! 💯

rdig reviewed (approved):
All good here! Nicely done!
